### ISSUE TITLE
Regularly save database snapshots

### ISSUE THREAD

**Issue by theoreticalbts** [steemit/steem#284]
Basically we want to bootstrap the undo DB infrastructure into being able to create snapshots.  For each snapshot, we want to have an iterator which walks through objects in the snapshot, possibly with concurrent modifications to the main DB.

To create a snapshot:
- Save the ID ranges for each object type.
- Keep undo states between current and snapshot.
- When querying for an ID, check each undo state and the current version; prefer oldest undo state.

To walk through the snapshot:
- Grab the next available ID from DB and lookaside, whichever is less will be returned.

Once we have this iterator, we'll be able to save a copy of the DB concurrently.  For example, every 5000 objects or 1MB, sleep for 10ms (unblocking network connections).  Even if new tx's / blocks come in, we can still save a faithful copy of the database state at the point in time of the snapshot, without having to pause long enough to iterate through the whole DB.

**Comment by theoreticalbts** [comment]
If snapshots are cheap, there's no reason not to keep a rolling snapshot every hour, which will allow much faster reloading of the chain.

**Comment by theoreticalbts** [comment]
Snapshots can be invalidated by schema changes or plugin changes #264 #285

**Comment by mvandeberg** [comment]
We have plans for snapshotting in the future and will need a detailed spec. Closing this in favor of a newer issue along with a spec when we get ready to implement this.
